fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        self.num_outputs = num_outputs
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_pipeline_record(self, users: List[Tuple[rpc.RRef, int, int]]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(self.device, self.rank, self.chunks, self.num_inputs, self.num_outputs, users)
+
+    def run(self, pipeline_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism. It modifies the given batches in place."""
+        m = len(pipeline_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for i in range(m):
+            with record_function("feed"):
+                pipeline_record.wait_for(i)
+            self.fence(pipeline_record, i)
+            self.compute(pipeline_record, i)
+            self.forward_results(i, pipeline_record)
+
+    def fence(self, pipeline_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Prepares micro-batches for computation."""
+        # Ensure that batches[chunk-1] is executed after batches[chunk] in
+        # backpropagation by an explicit dependency.
+        # TODO: This dependency injection causes deadlock if this partition
+        # gets its input from model input. 1) Figure out why 2) If we need to live
+        # with this constraint, replace the condition 'pipeline_record.rank > 0' below with
+        # a more accurate one.
+        if chunk != 0 and pipeline_record.users and pipeline_record.rank > 0:
+            dependant_tensors = []
+            batch = pipeline_record.batches[chunk]
+            assert batch is not None
+            for tensor, remote_ph_list in zip(batch.tensors, pipeline_record.forwarded_phony[chunk - 1]):
+                dependant = tensor
+                for remote_ph in remote_ph_list:
+                    ph = remote_ph.to_here()
+                    dependant = join(dependant, ph)
+                dependant_tensors.append(dependant)
+            pipeline_record.batches[chunk] = Batch(tuple(dependant_tensors), chunk)
+
+    def compute(self, pipeline_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Runs tasks with synchronization to tensor-pipe streams."""
+        checkpoint_stop = self.checkpoint_stop
+
+        # Disable checkpointing if in eval mode.
+        if not self.module.training:
+            checkpoint_stop = 0
+
+        exc_info: Optional[ExcInfo] = None
+
+        batch = pipeline_record.get_batch(chunk)
+
+        if pipeline_record is not None and pipeline_record.rank >= 0:
+            for e in pipeline_record.recv_events[chunk]:
+                if e is not None and is_cuda(self.stream):
+                    self.stream.wait_event(e)
```

Comment:
I think the code would be easier to reason about and more modular if this logic was pulled into PipelineRecord.get_batch.

To understand this snippet of code I have to return back PipelineRecord. Modularity and layering would greatly aid readability.

Reply:
I refactored this part to a separate method of PipelineRecord. get_batch is a light weight getter and is used elsewhere also.